app-ng/src/job.lisp

```diff
@@ -0,0 +1,145 @@
+
+(define-condition job-error (error)
+  ((job :initarg :job :reader job-error-job)
+   (message :initarg :message :reader job-error-msg))
+  (:report (lambda (c s)
+             (write-string (job-error-msg c) s)))
+  (:documentation "Error signaled because of job mismanagement."))
+
+(defmacro with-job-lock ((job &key (wait t)) &body body)
+  "Evaluate BODY with the LOCK acquired for the duration of BODY. If WAIT is T, block until the lock is available."
+  (alexandria:once-only (job wait)
+    `(when (bt:acquire-lock (job-lock ,job) ,wait)
+       (unwind-protect
+            (progn ,@body)
+         (bt:release-lock (job-lock ,job))))))
+
+(defmethod print-object ((job job) stream)
+  (print-unreadable-object (job stream :type nil :identity t)
+    (format stream "QVM Job (~A)" (job-status-name job))))
+
+;; The %x-unsafe variants below do not acquire a lock.
+
+(defun %job-running-p-unsafe (job)
+  (adt:match job-status (job-status job)
+    (job-status-running t)
+    (_ nil)))
+
+(defun job-running-p (job)
+  "Is the JOB in the RUNNING state."
+  (with-job-lock (job)
+    (%job-running-p-unsafe job)))
+
+(defun %job-killed-p-unsafe (job)
+  (adt:match job-status (job-status job)
+    (job-status-killed t)
+    (_ nil)))
+
+(defun job-killed-p (job)
+  "Is the JOB in the KILLED state?"
+  (with-job-lock (job)
+    (%job-killed-p-unsafe job)))
+
+(defun %job-finished-p-unsafe (job)
+  (adt:match job-status (job-status job)
+    ((job-status-finished _) t)
+    (_ nil)))
+
+(defun job-finished-p (job)
+  "Is the JOB in the FINISHED state?"
+  (with-job-lock (job)
+    (%job-finished-p-unsafe job)))
+
+(defun make-job (fn)
+  "Make a JOB object which will, under JOB-START, run FN in a thread.
+
+The returned JOB can be repeatedly started if the JOB-STATUS is FINISHED (or KILLED). Only interact with the job after acquiring the lock (or alternatively by using WITH-JOB-LOCK)."
+  (%make-job (lambda (job)
+               (adt:match job-status (job-status job)
+                 (job-status-fresh
+                  (setf (job-thread job)
+                        (bt:make-thread
+                         (lambda ()
+                           (let ((result 
+                                   (handler-case
+                                       (job-status-finished (funcall fn))
+                                     (error (c)
+                                       (job-status-error c)))))
+                             (with-job-lock (job)
+                               (setf (job-status job) result))))))
+                  (setf (job-status job) job-status-running))
+                 (_ (error 'job-error :job job :message "Job cannot be started because it is not fresh."))))))
+
+(defmacro define-job-action (action (job) &body body)
+  "Evaluate BODY with a lock held for JOB."
+  (multiple-value-bind (body decls docstr)
+      (alexandria:parse-body body :documentation t)
+    (declare (ignore decls))
+    `(defun ,action (,job)
+       ,docstr
```

Comment:
`,@(alexandria:ensure-list docstr)`

Reply:
Why?

Reply:
I see that you can (technically) provide multiple docstrs to a defun. SBCL warns when you do so, and alexandria:parse-body will error if you do so.

Reply:
probably in case `docstr` is `nil`  ?

Reply:
What does that give you? `(defun a () nil 1)` is OK

Reply:
ok but ugly. at least, this is why I usually splice in docstring. don't want to put words in @stylewarning's mouf